support for :saveas {file} and :w {file}

sanchon opened the pull request:
After a bit of fiddling I think I've found a way to allow these two commands to work as expected. The trick is to compare the destination file name ("amatch") with the current buffer file name. If they are different then the user is trying to save the buffer contents to a different (new) file. 

:saveas was a bit more difficult, in this case the amatch and %:p have the same value (the new file) and the only way I found to detect what was happening was to use the BufFilePre event handler. 

I hope you like it!

m1foley commented on autoload/SimplenoteUtilities.py:
```diff
@@ -327,51 +329,165 @@ def display_note_in_scratch_buffer(self, note_id=None):
+
+
+        if userAction=="updatingNote":
+            #
```
Stray comment

sanchon replied:
fixed!

m1foley commented:
Thanks for doing this!

sanchon commented on autoload/SimplenoteUtilities.py:
```diff
@@ -327,51 +329,165 @@ def display_note_in_scratch_buffer(self, note_id=None):
+        if os.path.basename(currentfile) != os.path.basename(amatch):
+            userAction="writingBufferToADifferentFile" # user is executing :w <newfile>
+        else:
+            if renaming == "0":
```
fixed!

m1foley commented on autoload/SimplenoteUtilities.py:
```diff
@@ -327,51 +329,159 @@ def display_note_in_scratch_buffer(self, note_id=None):
+        else:
+            userAction="renamingBuffer"
+
+        if userAction=="updatingNote":
```
This is a really big `if` statement. I bet it would be easier to maintain if it were extracted into 3 separate methods (`update_note()`, etc.)

sanchon replied:
I was lazy... let's do some refactoring to make everything more mantainable

atomicules commented:
Hi, just to prove I've not completely forgotten about this:

Just run through the tests and I see some failures. That might be because the tests need updating to reflect reality. Not sure yet. I really do appreciate your efforts though, I'm just... busy.

Will take more of a look over this weekend.

atomicules commented:
Hi,

took more of a detailed look at this. The tests fail because this breaks `:SimplenoteUpdate`. I found that was because `amatch` is blank and so it [hits this first if statement](https://github.com/sanchon/simplenote.vim/blob/4510bf1fa41bd535b02b7bfe13c18d190f78f076/autoload/SimplenoteUtilities.py#L346) so I tried a fix like this:

`if (os.path.basename(currentfile) != os.path.basename(amatch)) and amatch:`

which solves that problem and allows `:Simplenote` update to work once. But because updating a note calls [`set_current_note`](https://github.com/sanchon/simplenote.vim/blob/4510bf1fa41bd535b02b7bfe13c18d190f78f076/autoload/SimplenoteUtilities.py#L402) this means the Vim [`:file`](https://github.com/sanchon/simplenote.vim/blob/4510bf1fa41bd535b02b7bfe13c18d190f78f076/autoload/SimplenoteUtilities.py#L36) command is called which means [`BufFilePre`](http://vimdoc.sourceforge.net/htmldoc/autocmd.html#BufFilePre) is triggered, which means the next time `:SimplenoteUpdate` is called [`s:renaming`](https://github.com/sanchon/simplenote.vim/blob/4510bf1fa41bd535b02b7bfe13c18d190f78f076/autoload/SimplenoteUtilities.py#L344) is then 1.

So right now I don't have a good fix for this, but I'll carry on having a think about it.

sanchon commented:
I fixed the problem with `:SimplenoteUpdate`. The `update_note_from_current_buffer` function has now a weird logic that tries to determine whether the user is executing `:w` or `:w {file}` or `:saveas {file}` and `:SimplenoteUpdate` doesn't need any of that. The old logic is kept now (intact) in a new function called `update_note_to_web_service` so the easiest fix is to call it directly.

I also fixed the tests and added a couple more for the new functionality.

atomicules commented:
@sanchon Thanks again for all this work. I reviewed it, wow two weeks ago now, but didn't get around to fixing one last glitch until now - sorry.

Are you happy with the tiny little change I've added just up there ^ for `update_note_from_current_buffer` ? If so I'll go-ahead and merge in.

atomicules commented:
Merged! Thanks! I'll spin up a new release shortly.
